dianatanase/assignment2/Q1.java

```diff
@@ -0,0 +1,159 @@
+import java.util.ArrayList;
+
+// Represent the Tree with Nodes that store data of generic type T.
+// Each node as a key, I assumed keys are positive integers and unique for each node.
+// Each node has a leaf and right child (could be null = leaf).
+// Store the parent of each node as well.
+class Node<T>{
+	
+	T data = null;
+	Node<T> left = null;
+	Node<T> right = null;
+	Node<T> parent = null;
+	int key = -1;
```

Comment:
As the key cannot (and mustn't) change; it is good practice to set it to be a final variable.
final int key;